Kinematics tutorial

sfmig opened the pull request:
Adding a kinematics tutorial!

I think I got a bit carried away, but hopefully it showcases well the possibilities ✨ 

## Description

**What is this PR**

- [ ] Bug fix
- [ ] Addition of a new feature
- [x] Other

**Why is this PR needed?**
We want a demo of the implemented kinematic functionalities.

**What does this PR do?**
Adds a notebook that renders in the sphinx docs, and shows how to compute the basic kinematic variables.

## References

Please reference any existing issues/PRs that relate to this PR.

## How has this PR been tested?

Please explain how any new code has been tested, and how you have ensured that no existing functionality has changed.

## Is this a breaking change?

If this PR breaks any existing functionality, please explain how and why.

## Does this PR require an update to the documentation?

If any features have changed, or have been added. Please explain how the
documentation has been updated.

## Checklist:

- [ ] The code has been tested locally
- [ ] Tests have been added to cover all new functionality
- [ ] The documentation has been updated to reflect any changes
- [ ] The code has been formatted with [pre-commit](https://pre-commit.com/)

niksirbi commented on examples/compute_kinematics.py:
```diff
@@ -0,0 +1,439 @@
+# ------------------------
+# First, we load an example dataset. In this case, we select the
+# `SLEAP_three-mice_Aeon_proofread` sample data.
+file_path = sample_data.fetch_sample_data_path(
+    "SLEAP_three-mice_Aeon_proofread.analysis.h5"
+)
```
You can leave this as is, or you may also use the more direct `fetch_sample_data` method. In the first example, I used the one with the path because I wanted to showcase loading from files. That said, reminding people how to load from a file is not a bad thing (even if its a bit of duplication).

niksirbi reviewed (changes requested):
Thanks a lot @sfmig!

Here's a summary of my most importand comments:
- I think it's better to showcase the accessor way of getting the kinematics, i.e. `ds.move.velocity` instead of `kin.compute_velocity(pose_tracks)`.
- Try to build the docs locally and you'll see the rendering errors I refer to in my comments. It's one type of error but occurs throughout the document
- I think this example is too long. I propose taking all the things about fitting circles (and comparing/plotting to ideal circles) out of it. This is really cool stuff, but it's kind of specific to this example. People who will click on this example will want to get to the "meat" (aka, how do I get the velocity) fast, and this additional stuff detracts from that (plus it introduces the extra `circle-fit` dependency). That said, we should definitely keep this version in a branch, because we could showcase is as a separate standalone example (or even implement some if it in movement).

Other than that, there's a bit of redundancy between this example and the first one I had written (e.g. visualising trajectories), but I propose we live with this redundancy for now and reorganise the examples intal smaller more atomic units later.

Oh, and I'd also name this example "Compute and visualise kinematics", as it's quite viz-seavy with the quiver plots etc.

niksirbi commented:
@sfmig I'd suggest rebasing to `main` before you proceed.
We recently merged some PRs that are relevant for this one:
- #142
- #143 

I've also updated the [section of the contributing guide about adding examples](https://movement.neuroinformatics.dev/community/contributing.html#updating-the-examples).

sfmig commented on examples/compute_kinematics.py:
```diff
@@ -0,0 +1,439 @@
+
+# %%
+# We can also color color the data based on its timestamp:
+fig, axes = plt.subplots(3, 1)
```
Noted, I removed a bit the clutter with `sharey=True`.

I only set y-axis to be shared to prioritise making the x-axis and y-axis scales equal, since the trajectory is expected to be circular. Setting both `sharex=True, sharey=True` is incompatible with `ax.axis("equal")`. If I only set `sharex=True` and axis equal, it crops out the data
![image](https://github.com/neuroinformatics-unit/movement/assets/33267254/630111ba-f534-45a1-aa2d-e6b2db173d37)

sfmig commented on examples/compute_kinematics.py:
```diff
@@ -0,0 +1,439 @@
+# uses second order accurate central differences.
+
+# %%
+# We can also visualise the speed, as the norm of the velocity vector:
```
I added [an issue](https://github.com/neuroinformatics-unit/movement/issues/147) to think about this

sfmig commented on examples/compute_kinematics.py:
```diff
@@ -0,0 +1,439 @@
+#  Therefore,
+# the opposite vector will point from the position point at ``t``, to the
+#  position point at ``t-1``
+# This opposite vector is what we represent in the plot.
```
I spelled this out a bit more, hopefully it is more clear now

sfmig commented:
thanks @niksirbi for the feedback!

I think I addressed all your points, but let me know if I missed something:
- changed to the accessor syntax,
- removed the circle fit example, and
- clarified the deal with the displacement vector.

Re the last point, I think this reveals that the definition of the displacement may be convenient for the developers (it's just the diff function for an array) but a bit cumbersome for the users. I would suggest to change the way we define it to point to the position at t+1 instead. Good thing is that it is an easy fix. I captured it in this issue for now #136, but happy to discuss.

niksirbi reviewed (approved):
Thanks a ton @sfmig!

I added a few pedantic points. 
I agree with your assessment of the `displacement` vector situation. That's partly why these examples are so beneficial, they make us realise things that don't make sense from the user's perspective. So I think we should solve #136 as you've suggested.

Other than that, I've one last point to make. Using the `move` accessor is not just for convenience, it has an important side-effect which we haven't mentioned. Sorry if
I didn't properly communicate that earlier. I propose appending the following section at the end of this example:

```python
# %%
# .. _target-access-kinematics:
#
# Accessing pre-computed kinematic variables
# ------------------------------------------
# Once each kinematic variable has been computed via the ``move`` accessor,
# e.g. by calling ``ds.move.velocity``, the resulting data array will
# also be available as a dataset property, e.g. ``ds.velocity``.
# Since we've already computed ``displacement``, ``velocity`` and
# ``acceleration`` above, they should be listed among the data arrays
# contained in the dataset:
print(ds)

# %%
# Indeed we see that in addition to the original data arrays ``pose_tracks``
# and ``confidence``, the dataset now also contains data arrays called
# ``displacement``, ``velocity`` and ``acceleration``.

print(ds.displacement)

# %%
print(ds.velocity)

# %%
print(ds.acceleration)

```

We can cross-reference it earlier in the example in this way:
```python
# %%
# However, we encourage our users to familiarise themselves with the more
# convenient syntax of the ``move`` accessor. Another advantage of using the
# ``move`` accessor is that the computed data array will be automatically
# added to the dataset (see :ref:`target-access-kinematics`).
```

I've tested the cross-referencing locally and it works.

We can still have future discussions about whether this behaviour of the accessor makes sense, but I thought we should mention it in the example already (since that is what `movement` currently does).

sfmig commented on examples/compute_kinematics.py:
```diff
@@ -0,0 +1,394 @@
+# %matplotlib widget
+from matplotlib import pyplot as plt
+
+import movement.analysis.kinematics as kin
```
found a way to do this And keep ruff happy :)
